decidim/decidim
# Implement authorization data recovery for deleted users

#### :tophat: What? Why?
Implements a recovery procedure for data mapped to a conflicting authorization record for a deleted user as described in #9179.

This procedure will recover all data that can have authorization permission requirements on them but it won't check whether each component/record has been configured to require authorization for the authorizations to be recovered because of few reasons:
1. The permissions of each component can change at any given moment which is why we cannot assume that the recovery should be only done for records that currently require authorization
2. There is no data stored in the database that would allow us to check afterwards whether the record was originally created based on some existing authorization (because of point 1)

So for example, the budgeting component could allow the user casting a vote without an authorization. Later, the admin could notice that they forgot to set the authorization rules to the component and they fix this situation for the second day of voting. Now the user already cast a vote which did not require any authorization but any further votes will require authorization. We do not store any information that would allow the budgeting vote/order itself to tell whether it required authorization or not when it was cast.

Therefore, to avoid unnecessary complexity with the transfer procedures, it is assumed that all records that may or may not require an authorization needs to be transferred over to the new account.

This PR also introduces the `AuthorizationTransfer` and `AuthorizationTransferRecord` models into the core in order to trace back the transfers and store information about the transferred records. The amount of each transferred record with the record type is showed to the user when they authorize their account and data was transferred (see the screenshots).

It is also possible to disable the transfer functionality completely or disable the record transfers for a specific module. See the added CHANGELOG entries for explanation of the feature and further information about the changes.

#### :pushpin: Related Issues
- Fixes #9179

#### Testing
- Login with a regular user, e.g. `user@example.org`
- Authorize this user with the example authorization, take note for the "document number" you used to be able to provide the same number later on
- Go to one of the budgets component and cast a vote there
- After the vote is cast, delete the account using the account delete form
- Re-register on the platform, it does not matter what data you submit in the registration form
- Go to the same budgets component and see that there is currently no budget vote in the component
- Authorize the new user with a matching `unique_id` (i.e. the "document number") using the example authorization
- Go back to the budgets component and see that the previous vote was correctly transferred to the new user

#### :clipboard: Checklist


### :camera: Screenshots
Before deleting the account that has an authorization, the following extra message will be displayed at the delete account page (visible **only** when the user has authorizations):

![Delete account message](https://user-images.githubusercontent.com/864340/178319415-103270f4-9bee-45e8-a23a-cc9446145cd3.png)

After a successful authorization which was transferred from a deleted user, the user will see the amount of each transferred record type in the flash message:

![Successful authorization transfer](https://user-images.githubusercontent.com/864340/178296308-5b4928dd-adc3-425f-8d33-fc27c0bc35ad.png)
